module3-crud-and-data-models/r1-schema-and-data-models/README.md

```diff
@@ -0,0 +1,161 @@
+
+### What are database schemas?
+
+A database schema is a `blueprint` or `architecture` of how our data will look. It doesn’t hold data itself, but instead describes the shape of the data and how it might relate to other tables or models.
+
+For example, This is a `Customer` schema in mongoDB using mongoose.
+
+```js
+const mongoose = require("mongoose");
+
+const customer = new mongoose.Schema({
+  name: {
+    type: String,
+    required: true,
+  },
+  zipcode: {
+    type: Number,
+  },
+});
+
+module.exports = mongoose.model("Customer", customer);
+```
+
+The same `Customer` schema in a MySql database.
+
+```sql
+CREATE TABLE customer (
+ id INT AUTO_INCREMENT PRIMARY KEY,
+ name VARCHAR(50) NOT NULL,
+ zipcode INT(15) default NULL,
+)
+
+```
+
+We can see that the schema definition for the `Customer` has a `name` which is a type `string` and a `zipcode` that is of type `number`.
+
+### The Difference between data model and schema
+
+The database schema is one that contains list of attributes and instructions to tell the database engine how data is organised whereas Data model is a collection of conceptional tools for describing data, data-relationship and consistency constraints.
+
+## Introduction to CRUD operations
+
+When we are building APIs, we want our models to provide four basic types of functionality. The model must be able to Create, Read, Update, and Delete resources. Computer scientists often refer to these functions by the acronym `CRUD`. A model should have the ability to perform at most these four functions in order to be complete.
+
+### CRUD and REST
+
+In a REST environment, CRUD often corresponds to the HTTP methods `POST`, `GET`, `PUT`, and `DELETE`, respectively. These are the fundamental elements of a persistent storage system.
+
+For example, imagine we are working with a system that is keeping track of meals and their corresponding prices for a restaurant. Let’s look at how we would implement CRUD operations.
+
+#### Create
+
+To create resources in a REST environment, we most commonly use the HTTP POST method. POST creates a new resource of the specified resource type.
+Imagine that we are adding a new food item to the stored list of dishes for this restaurant, and the dish objects are stored in a dishes resource.
+
+Request:
+`POST http://www.myrestaurant.com/api/dishes/`
+
+We will need to send the `dish` data too.
+
+Body -
+
+```json
+{
+  "dish": {
+    "name": "Avocado Toast",
+    "price": 8
+  }
+}
+```
+
+Response: Status Code - 201 (Created)
+
+This creates a new item with a `name` value of `“Avocado Toast”` and a price value of 8. Upon successful creation, the server should return a header with a link to the newly-created resource.
+
+#### Read
+
+To read resources in a REST environment, we use the GET method. Reading a resource should never change any information - it should only retrieve it.
```

Comment:
you should explain that this is not strictly enforced in code but by convention, and also i think in general you need more explaining that REST is a convention. you can technically do whatever you want but it would be confusing to users of your API